File: Java/Array/UniqueSubarray.java
/*
You are given an integer array nums.
You are allowed to delete any number of elements from nums without making it empty. 
After performing the deletions, select a subarray of nums such that:

All elements in the subarray are unique.
The sum of the elements in the subarray is maximized.
Return the maximum sum of such a subarray. 
*/

import java.util.HashSet;
import java.util.Set;

public class UniqueSubarray{
    public static int maxSum(int[] arr){
        Set<Integer> set = new HashSet<>();
        boolean allNeg = true;
        int maxValue = Integer.MIN_VALUE, currSum = 0;
        
        for(int i : arr){
            if(i >= 0){
                allNeg = false;
            }
            if(i > maxValue){
                maxValue = i;
            }
        }
        if(allNeg) return maxValue;

        for(int i : arr){
            set.add(i);
        }
        for(int val : set){
            if(val > 0) currSum += val;
        }
        return currSum;
    }
    public static void main(String... args){
        int[] arr = {1, 2, 3, 4 ,5};
        System.out.println(UniqueSubarray.maxSum(arr));
    }
}

File: Golang/Array/UniqueSubarray.go
/*You are given an integer array nums.

You are allowed to delete any number of elements from nums without making it empty. 
After performing the deletions, select a subarray of nums such that:

All elements in the subarray are unique.
The sum of the elements in the subarray is maximized.
Return the maximum sum of such a subarray.*/
// package main
// import (
// 	"sort"
// 	"fmt"
// )

// func maxSum(arr []int)int{
// 	sort.Ints(arr)
// 	prev := arr[len(arr) - 1]
// 	sum := prev

// 	for i := len(arr) - 2; i >= 0; i--{
// 		n := arr[i]
// 		if n < 0{
// 			return sum
// 		}else if n != prev{
// 			sum += n
// 		}
// 		prev = n
// 	}
// 	return sum;
// }

// func main(){
// 	arr := []int{1, 2, 3, 4, 5}
// 	fmt.Println(maxSum(arr))
// }